index.js

```diff
@@ -16,46 +16,112 @@ class ExecutorRouter extends Executor {
      */
     constructor(config = {}) {
         const ecosystem = config.ecosystem || {};
-        const executorConfig = config.executor;
-        const defaultPlugin = config.defaultPlugin;
+        const { executor, defaultPlugin } = config;
 
-        if (!executorConfig || !(Array.isArray(executorConfig)) || executorConfig.length === 0) {
+        if (!executor || !Array.isArray(executor) || executor.length === 0) {
             throw new Error('No executor config passed in.');
         }
-
         super();
 
-        executorConfig.forEach((plugin) => {
-            let ExecutorPlugin;
+        let ExecutorPlugin;
+
+        this._executors = [];
 
+        executor.forEach(plugin => {
             try {
                 // eslint-disable-next-line global-require, import/no-dynamic-require
                 ExecutorPlugin = require(`screwdriver-executor-${plugin.name}`);
+                this._executors.push(plugin);
             } catch (err) {
-                console.error(err);
+                console.error(err.message);
 
                 return;
             }
+            // Add ecosystem to executor options
+            const options = Object.assign({ ecosystem }, plugin.options);
 
-            const options = Object.assign({ ecosystem }, plugin.options); // Add ecosystem to executor options
-            const executorPlugin = new ExecutorPlugin(options);
+            this[plugin.name] = new ExecutorPlugin(options);
+        });
 
-            if (defaultPlugin && defaultPlugin === plugin.name) {
-                // set the default plugin to the desired one
-                this.defaultExecutor = executorPlugin;
-            } else if (!defaultPlugin && !this.defaultExecutor) {
-                // When given no default plugin, use the first module we instantiated
-                this.defaultExecutor = executorPlugin;
+        // executor rules chain
+        this._executorRules = [
+            {
+                name: 'annotated',
+                priority: 1,
+                check: buildConfig => {
+                    const annotations = this.parseAnnotations(buildConfig.annotations || {});
+
+                    return annotations[ANNOTATION_EXECUTOR_TYPE];
+                }
+            },
+            {
+                name: 'weighted',
+                priority: 2,
+                check: () => this.getWeightedExecutor(this._executors)
+            },
+            {
+                name: 'default',
+                priority: 3,
+                check: () => defaultPlugin || (this._executors[0] && this._executors[0].name)
             }
+        ];
+        // sort by priority and store
+        this._executorRules.sort((a, b) => a.priority - b.priority);
```

Comment:
why do we need to sort here isn't order guaranteed to be insertion order ?

Reply:
yeah order is preserved, until code is moved around. Probably felt this check would be safe.